### PR TITLE
Prometheus Ceph Exporter

### PR DESCRIPTION
**Pull request by kfox1111**
Adds Prometheus support for gathering metrics from Ceph.

### CODE REVIEW

**Comment by prydonius** [comment]
ping @mgoodness

**Comment by kfox1111** [comment]
ping

File: stable/ceph-exporter/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: ceph-exporter
+version: 1.0.0
```
Comment (unguiculus) [inline stable/ceph-exporter/Chart.yaml:2]: Add `appVersion`.

File: stable/ceph-exporter/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+description: Prometheus ceph exporter module
+keywords:
+- prometheus
+- ceph
+maintainers:
+  - name: Kevin Fox
```
Comment (unguiculus) [inline stable/ceph-exporter/Chart.yaml]: Use Github username.

File: stable/ceph-exporter/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{template "fullname" .}}
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/deployment.yaml]: Use consistent formatting throughout the chart:
Space after `{{` and before `}}`.

File: stable/ceph-exporter/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    app: {{ template "fullname" . }}
+    heritage: "{{ .Release.Service }}"
+    release: "{{ .Release.Name }}"
+    chart: "{{.Chart.Name}}-{{.Chart.Version}}"
+spec:
+  replicas: {{default 1 .Values.replicaCount}}
+  template:
+    metadata:
+      labels:
+        app: {{template "fullname" .}}
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/deployment.yaml]: Please apply best practices for labels: https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/labels.md
Note that you will also have to update label selectors.

See chart examples for reference: https://github.com/kubernetes/helm/tree/master/docs/examples

File: stable/ceph-exporter/templates/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      annotations:
+        "prometheus.io/port": "9128"
+        "prometheus.io/scrape": "true"
+    spec:
+      hostNetwork: True
+      nodeSelector:
+{{ toYaml .Values.nodeSelector | indent 10 }}
+      containers:
+      - name: {{ template "fullname" . }}
+        image: "{{default "digitalocean/ceph_exporter" .Values.image}}:{{default "1.0.0" .Values.imageTag}}"
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/deployment.yaml]: Add default values to `values.yaml`. Then you won't have to specify them here. Apply everywhere.

File: stable/ceph-exporter/values.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# Default values for ceph-exporter.
+# This is a YAML-formatted file.
+# Declare name/value pairs to be passed into your templates.
+# name: value
+
+Name: ceph-exporter
```
Comment (unguiculus) [inline stable/ceph-exporter/values.yaml]: Values should start with a lower case letter.

File: stable/ceph-exporter/values.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+replicaCount: 1
+
+nodeSelector: {}
+
+ServiceType: ClusterIP
+
+configmap: ceph-read-only
+secret: ceph-read-only
+
+resources:
```
Comment (unguiculus) [inline stable/ceph-exporter/values.yaml]: We usually recommend not to specify defaults for resources.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for your PR. Sorry for the late feedback.

**Comment by kfox1111** [comment]
@unguiculus thanks for the review. I think I fixed all the issues raised.

File: stable/ceph-exporter/templates/deployment.yaml [hunk 8]
```diff
@@ -1,19 +1,21 @@
...
 spec:
-  replicas: {{default 1 .Values.replicaCount}}
+  replicas: {{ .Values.replicaCount }}
   template:
     metadata:
       labels:
-        app: {{template "fullname" .}}
-        release: {{.Release.Name | quote }}
+        app: {{ template "name" . }}
+        heritage: "{{ .Release.Service }}"
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/deployment.yaml]: Only `app` and `release` label should be here.

File: stable/ceph-exporter/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+spec:
+  replicas: {{ .Values.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "name" . }}
+        heritage: "{{ .Release.Service }}"
+        release: "{{ .Release.Name }}"
+        chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+      annotations:
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/deployment.yaml]: Make these annotations optional. They are not necessary when using the Prometheus Operator.

File: stable/ceph-exporter/templates/_helpers.tpl [hunk 10]
```diff
@@ -0,0 +1,17 @@
+{{/* vim: set filetype=mustache: */}}
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/_helpers.tpl:1]: Fix formatting issues around braces and truncate to 63 chars. Ideally use `helm create` with Helm 2.5 to get a correct `_helpers.tpl`.

File: stable/ceph-exporter/templates/deployment.yaml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  labels:
+    app: {{ template "name" . }}
+    heritage: "{{ .Release.Service }}"
+    release: "{{ .Release.Name }}"
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+spec:
+  replicas: {{ .Values.replicaCount }}
+  template:
+    metadata:
+      labels:
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/deployment.yaml:15]: Only `app` and `release` label should be used here.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/ceph-exporter/templates/deployment.yaml (hunk 9 above)
Comment (kfox1111) [reply stable/ceph-exporter/templates/deployment.yaml]: how does prometheus operator figure out what pods to scrape? I'm confused.

File: stable/ceph-exporter/templates/svc.yaml [hunk 12]
```diff
@@ -0,0 +1,19 @@
+apiVersion: v1
```
Comment (unguiculus) [inline stable/ceph-exporter/templates/svc.yaml]: Rename the file to `service.yaml`.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please add a readme file.

**Comment by unguiculus** [comment]
The readme should have a table with config options.

**Comment by unguiculus** [comment]
@k8s-bot ok to test

**Comment by kfox1111** [comment]
@unguiculus  I mentioned in another review, a lot of the charts have large amounts of documentation and duplicate stuff (a source of bugs). Should we have a "how to use helm charts" document instead of making every chart carry the load? All the settable values are inside the values.yaml file. Why do we need to duplicate them and then have to keep in sync every option available in both files?

**Comment by kfox1111** [comment]
Perhaps we should instead have a 'helm describe-values' or something feature that presents all the values in a helm chart to the user in a pretty way?

**Comment by kfox1111** [comment]
So, question. how do you have a chart that must have external secret/external storage to ever work with the e2e tests? Can we disable them per chart?

**Comment by unguiculus** [comment]
The usual way is to also provide an option that the secrets be created by the chart. The chart should install successfully with defaults if possible. That being said, please provide such an option.

**Comment by kfox1111** [comment]
The format of the secret is pretty particular. It would be much more effort for the user to get the secret passed through values then just kubectl create secret....

I've been a little leary of putting secrets in helm charts too. It puts them in configmaps and other places that may not be the more secure place for them?

And third, passing it through the chart still wont let the container start as it won't have a ceph in e2e? Or is it just checking the secrets/configmaps all exist, not that the pod can actually hits a running state?

**Comment by unguiculus** [comment]
The tests check that the pods hit running state. One option would be to make it fail gracefully. You can look at the Datadog charts as an example which requires an API key.

https://github.com/kubernetes/charts/blob/master/stable/datadog/templates/deployment.yaml#L2
https://github.com/kubernetes/charts/blob/master/stable/datadog/templates/NOTES.txt

**Comment by kfox1111** [comment]
hmm... I guess that could work. still feels a little weird having the chart successfully deploy of the box to appease the tests but be otherwise unusable.

I guess it does give a nice place to print out notes to tell the user how to update it...

Should the notes in README be moved to NOTES then?

**Comment by unguiculus** [comment]
I guess I'd do it like so:
* Document it in the README. 
* Add a prominent message to NOTES referring to the documentation in the README.

What do you think?

**Comment by kfox1111** [comment]
The thing I like about notes, is it shows with packages.

helm install stable/X, then you see the notes.

with readme, you have to checkout a git repo and go hunt for it, or download the package tarball and extract it to get the readme.

**Comment by unguiculus** [comment]
Well, we kind of need both. You see the Readme on Github and on kubeapps.com *before* you install the chart.

**Comment by kfox1111** [comment]
I agree this is a problem. And a furtherance of the huge amount of duplication in each chart, either within each chart or across charts. :/

We might want to put our collective heads together and hash through some of this?

Maybe readme and/or notes should be automatically generated somehow?

**Comment by prydonius** [comment]
I think this discussion raises some good points about the duplication in chart READMEs/NOTES. @kfox1111 it would be great to discuss this further in a future Charts dev call (every Thus 11am PT) or SIG-Apps as we go to plan priorities for 1.8.

On the successful install by default requirement, we did talk about having a test-values.yaml file that would be used by the CI for tests. This could be the place for fake API keys that allow the charts to be testable in CI, whilst allowing the chart developer to provide a more expected UX for users installing the chart. Another thing we can discuss during one of these next calls.

In the meantime, I think we can copy the approaches in the existing charts to have this PR go through. Once we have a better story for this, we can go back and update all the charts.

**Comment by kfox1111** [comment]
I should be able to make the 11am charts dev call I think. is there location info somewhere?

Ok. I'll see about fixing it up for now.

**Comment by kfox1111** [comment]
Ok. I still have the notes bit to add, but please review on the approach taken to add ceph configmap and secret generation. I expect to do a chart for https://github.com/kubernetes-incubator/external-storage/tree/master/ceph/rbd soon and it will reuse the configmap/secret charts the same way its used here.

**Comment by kfox1111** [comment]
Ok. I think this has all the things added.

**Review by mgoodness — COMMENTED** [review]
I'm going to defer to @unguiculus, as this PR has waited on me for far too long. Sorry!

**Comment by unguiculus** [comment]
If I remember correctly, this PR started out with one chart. Now, we have three individual charts. Can you explain why you made this change? It doesn't make sense to me because they don't seem to make sense on their own. Why not create one chart with multiple components? The Nginx Ingress Controller would be a good example for such a chart. See https://github.com/kubernetes/charts/tree/master/stable/nginx-ingress/.

**Comment by kfox1111** [comment]
It started off as one chart, but folks wanted "default" secrets and "default" configmap generation support added in. so I added those as separate charts so that they can be reused by other charts that are ceph related and will get the same kind of request in for. I have another one I have been working on to add ceph dynamic pvc support here: https://github.com/kfox1111/charts/tree/ceph-provisioner/stable/ceph-provisioner I will submit it soon, once I can verify it works properly.

**Comment by viglesiasce** [comment]
The pattern we have seen in other charts is to have a flag to provide your own "custom" configmap (or secret). 

Can we move to that pattern for both secrets and configmaps in the ceph chart?

https://github.com/kubernetes/charts/tree/master/stable/jenkins#custom-configmap

**Comment by viglesiasce** [comment]
Marking stale.

**Comment by viglesiasce** [comment]
Closing as stale.

**Pull request closed without merging**